features/apimgt/org.wso2.carbon.apimgt.admin.feature/src/main/resources/admin/source/src/app/components/Throttling/Subscription/AddEdit.jsx

```diff
@@ -215,6 +224,7 @@ function AddEdit(props) {
                     unitTimeEdit = result.body.defaultLimit.bandwidth.unitTime;
                     typeEdit = result.body.defaultLimit.type;
                 }
+                setValidRoles(result.body.permissions.roles);
```

Comment:
We can use the same dispatcher to handle the roles without introducing the new state variables ( validRoles, invalidRoles and roleValidity). Use the same error handling method to show errors or disable the submit button.  Inside the formSaveCallback method, you can format the payload required to send to the backend.

Reply:
Let's go with the current implementation. It seems like getting rid of the extra state variables is too much work and doesn't add value except removing an extra rerender at the initial component mount.